File: src/test/scala/com/exasol/cloudetl/source/ParquetSourceTestTest.scala
package com.exasol.cloudetl.source

import java.nio.file.Paths

import com.exasol.cloudetl.filesystem.FileSystemManager
import com.exasol.parquetio.reader.RowParquetReader

import org.apache.hadoop.fs.Path
import org.apache.parquet.hadoop.util.HadoopInputFile
import org.apache.parquet.io.InputFile
import org.apache.parquet.schema.MessageTypeParser

class ParquetSourceTestTest extends AbstractSourceTest {

  override val format: String = "parquet"

  test("getSchema returns parquet schema") {
    val expectedMessageType = MessageTypeParser
      .parseMessageType("""message spark_schema {
                          |  optional int64 sales_id;
                          |  optional int32 position_id;
                          |  optional int32 article_id;
                          |  optional int32 amount;
                          |  optional double price;
                          |  optional int32 voucher_id;
                          |  optional boolean canceled;
                          |}
      """.stripMargin)

    val filePattern = Paths.get(s"$resourceDir/sales_pos*.parquet")
    val globbedFilePath = FileSystemManager(getFileSystem()).getLocalFiles(filePattern)
    globbedFilePath.foreach { file =>
      val schema = RowParquetReader.getSchema(getInputFile(file))
      assert(schema === expectedMessageType)
    }
  }

  private[this] def getInputFile(path: Path): InputFile =
    HadoopInputFile.fromPath(path, getConf())

  test("stream returns count of records from single PARQUET file") {
    val filePath = Paths.get(s"$resourceDir/sales_positions1.snappy.parquet")
    assert(getRecordsCount(filePath) === 500)
  }

  test("stream returns count of records from PARQUET files") {
    val filePattern = Paths.get(s"$resourceDir/sales_positions*.parquet")
    assert(getRecordsCount(filePattern) === 1005)
  }

  test("stream returns count of records from PARQUET files with richer types") {
    val filePath = Paths.get(s"$resourceDir/sales1.snappy.parquet")
    assert(getRecordsCount(filePath) === 999)
  }

  test("stream throws if it cannot create PARQUET reader") {
    val nonPath = new org.apache.hadoop.fs.Path(s"$resourceDir/notFile.parquet")
    val thrown = intercept[java.io.FileNotFoundException] {
      getSource(nonPath).stream().size
    }
    assert(thrown.getMessage === s"File $nonPath does not exist")
  }

}

File: src/test/java/com/exasol/cloudetl/source/ParquetSourceTestTest.java
package com.exasol.cloudetl.source;

import static org.junit.jupiter.api.Assertions.*;

import java.nio.file.Paths;

import org.apache.hadoop.fs.Path;
import org.apache.parquet.hadoop.util.HadoopInputFile;
import org.apache.parquet.io.InputFile;
import org.apache.parquet.schema.MessageTypeParser;
import org.junit.jupiter.api.Test;

import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.filesystem.FileSystemManager;
import com.exasol.parquetio.reader.RowParquetReader;

class ParquetSourceTestTest extends AbstractSourceTest {
    @Override
    protected String getFormat() {
        return "parquet";
    }

    @Test
    void getSchemaReturnsParquetSchema() throws Exception {
        final var expectedMessageType = MessageTypeParser.parseMessageType(("message spark_schema {\n" +
"  optional int64 sales_id;\n" +
"  optional int32 position_id;\n" +
"  optional int32 article_id;\n" +
"  optional int32 amount;\n" +
"  optional double price;\n" +
"  optional int32 voucher_id;\n" +
"  optional boolean canceled;\n" +
"}\n" +
"\n"));

        final java.nio.file.Path filePattern = Paths.get(this.resourceDir + "/sales_pos*.parquet");
        for (final Path file : ScalaConverters
                .asJavaList(new FileSystemManager(getFileSystem()).getLocalFiles(filePattern))) {
            assertEquals(expectedMessageType, RowParquetReader.getSchema(getInputFile(file)));
        }
    }

    @Test
    void streamReturnsCountOfRecordsFromSingleParquetFile() throws java.io.IOException {
        assertEquals(500, getRecordsCount(Paths.get(this.resourceDir + "/sales_positions1.snappy.parquet")));
    }

    @Test
    void streamReturnsCountOfRecordsFromParquetFiles() throws java.io.IOException {
        assertEquals(1005, getRecordsCount(Paths.get(this.resourceDir + "/sales_positions*.parquet")));
    }

    @Test
    void streamReturnsCountOfRecordsFromParquetFilesWithRicherTypes() throws java.io.IOException {
        assertEquals(999, getRecordsCount(Paths.get(this.resourceDir + "/sales1.snappy.parquet")));
    }

    @Test
    void streamThrowsIfItCannotCreateParquetReader() {
        final Path nonPath = new Path(this.resourceDir + "/notFile.parquet");
        final IllegalStateException thrown = assertThrows(IllegalStateException.class,
                () -> getSource(nonPath).stream().size());
        assertInstanceOf(java.io.FileNotFoundException.class, thrown.getCause());
        assertEquals("File " + nonPath + " does not exist", thrown.getCause().getMessage());
    }

    private InputFile getInputFile(final Path path) throws java.io.IOException {
        return HadoopInputFile.fromPath(path, getConf());
    }
}
